bokehjs/src/coffee/document.coffee

```diff
@@ -650,8 +651,13 @@ export class Document
             throw new Error("Cannot apply patch to #{patched_id} which is not in the document")
           patched_obj = @_all_models[patched_id]
           attr = event_json['attr']
-          value = Document._resolve_refs(event_json['new'], old_references, new_references)
-          patched_obj.setv({ "#{attr}" : value })
+          model_type = event_json['model']['type']
+          if attr == 'data' and model_type == 'ColumnDataSource'
+            [data, shapes] = decode_column_data(event_json['new'])
+            patched_obj.setv({ "_shapes": shapes, "data" : data}, {notify: false})
```

Comment:
>What's the difference between notify=false and silent=true?
  
``silent=true`` completely skips telling the document about the change, which means models won't be invalidated and therefore not re-rendered. ``notify=false`` ensures the models are invalidated (and therefore re-rendered) but skips notifying any document callbacks, which means the ``ClientSession`` document listener isn't called and no ModelChangeEvent is generated. This significantly improves performance even independent of the other changes in this PR because when the client receives a ModelChangeEvent it no longer generates another ``ModelChangeEvent``, which has to be compared against the current data before it is invalidated by ``ClientSession._should_suppress_on_change``.